### PR TITLE
[build] strip the binary size

### LINKED ISSUE #1734: [release] Strip the binary size
**Summary**

https://github.com/johnthagen/min-sized-rust
we can strip the binary size for docker or releases.

### PR DESCRIPTION
**Pull request by PsiACE**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

for x86_64-unknown-linux-gnu
- reduce 90+% size for release
- reduce ~70% size for debug

## Changelog

- Improvement
- Build/Testing/CI

## Related Issues

Fixes #1734 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by PsiACE** [comment]
## after strip

**debug**

```
❯ bloaty databend-query -n 10
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  25.4%  48.9Mi  70.9%  48.9Mi    .text
  23.6%  45.5Mi   0.0%       0    .debug_info
  12.4%  23.9Mi   0.0%       0    .strtab
   8.1%  15.6Mi   0.0%       0    .debug_pubtypes
   6.3%  12.1Mi  11.9%  8.17Mi    [44 Others]
   5.7%  11.0Mi   0.0%       0    .debug_str
   4.7%  9.00Mi   0.0%       0    .debug_pubnames
   4.3%  8.23Mi   0.0%       0    .debug_line
   3.5%  6.72Mi   0.0%       0    .symtab
   3.4%  6.49Mi   9.4%  6.49Mi    .eh_frame
   2.8%  5.37Mi   7.8%  5.37Mi    .rela.dyn
 100.0%   192Mi 100.0%  68.9Mi    TOTAL
❯ bloaty databend-store -n 10
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  26.8%  40.4Mi  72.7%  40.4Mi    .text
  23.3%  35.1Mi   0.0%       0    .debug_info
  12.8%  19.3Mi   0.0%       0    .strtab
   7.5%  11.4Mi   0.0%       0    .debug_pubtypes
   6.3%  9.57Mi  11.8%  6.54Mi    [43 Others]
   5.6%  8.43Mi   0.0%       0    .debug_str
   4.3%  6.46Mi   0.0%       0    .debug_line
   4.0%  6.04Mi   0.0%       0    .debug_pubnames
   3.6%  5.49Mi   0.0%       0    .symtab
   3.5%  5.31Mi   9.6%  5.31Mi    .eh_frame
   2.2%  3.32Mi   6.0%  3.32Mi    .rela.dyn
 100.0%   150Mi 100.0%  55.6Mi    TOTAL
❯ bloaty databend-cli -n 10
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  41.3%  47.3Mi   0.0%       0    .debug_info
  14.3%  16.4Mi   0.0%       0    .debug_pubtypes
  10.1%  11.6Mi   0.0%       0    .debug_str
   8.3%  9.52Mi   0.0%       0    .debug_pubnames
   7.8%  8.96Mi  65.7%  8.96Mi    .text
   7.2%  8.20Mi   0.0%       0    .debug_line
   3.9%  4.45Mi   0.0%       0    .strtab
   3.7%  4.19Mi  16.2%  2.21Mi    [42 Others]
   1.3%  1.51Mi   0.0%       0    .symtab
   1.2%  1.36Mi   9.9%  1.36Mi    .eh_frame
   1.0%  1.11Mi   8.2%  1.11Mi    .rela.dyn
 100.0%   114Mi 100.0%  13.6Mi    TOTAL
```

**release**

```
❯ bloaty databend-query -n 10
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  19.0%  11.0Mi  50.1%  11.0Mi    .text
  14.6%  8.41Mi   0.0%       0    .strtab
  14.3%  8.25Mi   0.0%       0    .debug_str
  12.9%  7.46Mi   0.0%       0    .debug_info
   7.2%  4.17Mi   0.0%       0    .debug_pubnames
   7.1%  4.11Mi   0.0%       0    .debug_line
   6.9%  3.97Mi  12.5%  2.74Mi    [43 Others]
   5.6%  3.22Mi  14.7%  3.22Mi    .eh_frame
   5.1%  2.92Mi  13.3%  2.92Mi    .rela.dyn
   3.7%  2.12Mi   0.0%       0    .symtab
   3.6%  2.08Mi   9.5%  2.08Mi    .data.rel.ro
 100.0%  57.7Mi 100.0%  22.0Mi    TOTAL
❯ bloaty databend-store -n 10
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  20.4%  8.84Mi  53.8%  8.84Mi    .text
  14.4%  6.26Mi   0.0%       0    .strtab
  14.4%  6.25Mi   0.0%       0    .debug_str
  13.4%  5.80Mi   0.0%       0    .debug_info
   7.2%  3.14Mi   0.0%       0    .debug_line
   6.9%  2.98Mi   0.0%       0    .debug_pubnames
   6.0%  2.60Mi  10.3%  1.69Mi    [43 Others]
   5.4%  2.35Mi  14.3%  2.35Mi    .eh_frame
   4.9%  2.11Mi  12.8%  2.11Mi    .rodata
   3.6%  1.57Mi   0.0%       0    .symtab
   3.4%  1.46Mi   8.9%  1.46Mi    .rela.dyn
 100.0%  43.4Mi 100.0%  16.4Mi    TOTAL
❯ bloaty databend-cli -n 10
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  28.1%  8.72Mi   0.0%       0    .debug_str
  22.4%  6.94Mi   0.0%       0    .debug_info
  14.2%  4.41Mi   0.0%       0    .debug_pubnames
  13.4%  4.17Mi   0.0%       0    .debug_line
   6.4%  1.98Mi  43.5%  1.98Mi    .text
   4.0%  1.23Mi   0.0%       0    .strtab
   4.0%  1.23Mi  12.8%   597Ki    [41 Others]
   2.8%   901Ki  19.4%   901Ki    .rodata
   1.8%   567Ki  12.2%   567Ki    .rela.dyn
   1.8%   562Ki  12.1%   562Ki    .eh_frame
   1.2%   388Ki   0.0%       0    .symtab
 100.0%  31.1Mi 100.0%  4.55Mi    TOTAL
```

**Comment by BohuTANG** [comment]
How about the performance of the strip binary?

File: .cargo/config [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[target."x86_64-unknown-linux-gnu"]
+rustflags = ["-C", "link-arg=-Wl,--compress-debug-sections=zlib-gabi"]
```
Comment (sundy-li) [inline .cargo/config:2]: need newline

**Comment by ZhiHanZ** [comment]
> How about the performance of the strip databend-query binary?

I could do performance test to test on striped release if necessary

**Comment by PsiACE** [comment]
> How about the performance of the strip databend-query binary?

Needs testing, but I don't think this change will affect performance. The main thing eliminated is some debug-info.

**Comment by BohuTANG** [comment]
> > How about the performance of the strip databend-query binary?
> 
> Needs testing, but I don't think this change will affect performance. The main thing eliminated is some debug-info.

Cool, need a check.

**Comment by BohuTANG** [comment]
I think `opt-level = "z" ` is a performance killer, it explains in https://doc.rust-lang.org/cargo/reference/profiles.html:
"z": optimize for binary size, but also turn off loop vectorization.

And, I have a test
SQL:
SELECT max(number), sum(number) FROM numbers_mt(1000000000) GROUP BY number % 3, number % 4, number % 5;

The master version:
```
60 rows in set. Elapsed: 2.165 sec. Processed 1.00 billion rows, 8.00 GB (461.85 million rows/s., 3.69 GB/s.)
```

The strip version:
```
60 rows in set. Elapsed: 5.314 sec. Processed 1.00 billion rows, 8.00 GB (188.17 million rows/s., 1.51 GB/s.)
```

**Comment by BohuTANG** [comment]
Here we can remove `opt-level = "z"`, the size:
```
61M	target/release/databend-query-this-patch-without-optlevel
694M	target/release/databend-query-master
```
And the performance as fast as the master.

**Comment by PsiACE** [comment]
> Here we can remove `opt-level = "z"`, the size:

OK, I'll remove it. But for now the main task is to fix stateless-test.

**Comment by BohuTANG** [comment]
/lgtm

Very cool, keep the performance and gain the smaller binary(release) size:
This patch:
61M	target/release/databend-query

Master:
694M	target/release/databend-query

**Pull request merged**